src/compiler/html/inline-esm-import.ts

```diff
@@ -56,50 +74,85 @@ export const optimizeEsmImport = async (
+    const newPath = updateImportPathDir(orgImportPath, newDir);
     if (newPath) {
-      code = code.replace(`"${orgImportPath}"`, `"${newPath}"`);
-      code = code.replace(`'${orgImportPath}'`, `'${newPath}'`);
+      code = code.replace(new RegExp(`"${orgImportPath}"`, 'g'), `"${newPath}"`);
+      code = code.replace(new RegExp(`'${orgImportPath}'`, 'g'), `'${newPath}'`);
     }
-  });
+  }
 
   return {
     code,
-    orgImportPaths,
+    orgImportPaths: orgModulePaths,
   };
 };
 
-const replacePathDir = (orgImportPath: string, newDir: string) => {
+/**
+ * Update the directory of an ESM module specifier to include a new directory,
+ * e.g. by transforming `./foo.js` to `/build/foo.js`.
+ *
+ * @param orgImportPath the original path as found in the un-transformed source
+ * file
+ * @param newDir the new directory path which should be prepended to the
+ * original path
+ * @returns an updated path or `null`
+ */
+const updateImportPathDir = (orgImportPath: string, newDir: string): string | null => {
   if (orgImportPath.startsWith('./') && (orgImportPath.endsWith('.js') || orgImportPath.endsWith('.mjs'))) {
     return newDir + orgImportPath.substring(2);
   }
   return null;
 };
 
-const readImportPaths = (orgImportPaths: string[]): ts.TransformerFactory<ts.SourceFile> => {
-  return () => {
-    return (tsSourceFile) => {
-      const importStatements = tsSourceFile.statements
-        .filter(ts.isImportDeclaration)
-        .filter((s) => s.moduleSpecifier != null)
-        .filter((s) => ts.isStringLiteral(s.moduleSpecifier) && s.moduleSpecifier.text);
+/**
+ * Gather all module specifiers used in the `import` and `export` declarations
+ * in a bit of JS code
+ *
+ * @param code the code to transform
+ * @returns a list of the module specifiers present in the code
+ */
+function readModulePaths(code: string) {
+  const tsSourceFile = ts.createSourceFile('module.ts', code, ts.ScriptTarget.Latest);
+  const orgModulePaths: string[] = [];
 
-      importStatements.forEach((s) => {
-        if (ts.isStringLiteral(s.moduleSpecifier)) {
-          orgImportPaths.push(s.moduleSpecifier.text);
-        }
-      });
+  for (const stmt of tsSourceFile.statements) {
+    if (
+      isImportOrExportDecl(stmt) &&
+      stmt.moduleSpecifier != null &&
+      ts.isStringLiteral(stmt.moduleSpecifier) &&
+      stmt.moduleSpecifier.text
+    ) {
+      orgModulePaths.push(stmt.moduleSpecifier.text);
+    }
+  }
+  return orgModulePaths;
+}
 
-      return tsSourceFile;
-    };
-  };
-};
+/**
+ * Small type coercion / guard helper that returns whether or not a
+ * {@link ts.Statement} is an import / export declaration
+ *
+ * @param stmt the statement of interest
+ * @returns whether this is an import or export declaration or neither
+ */
+function isImportOrExportDecl(stmt: ts.Statement): stmt is ts.ImportDeclaration | ts.ExportDeclaration {
+  return ts.isExportDeclaration(stmt) || ts.isImportDeclaration(stmt);
+}
```

Comment:
this is I think necessary because I don't think TS is smart enough (yet!) to know that `ts.isExportDeclaration(foo) || ts.isImportDeclaration(foo)` evaluating to `true` means it must be one of those two types - instead when I inlined the usage of these two functions it was inferring a type of `ts.Statement` :/